Formula/clickhouse-odbc.rb

```diff
@@ -0,0 +1,103 @@
+class ClickhouseOdbc < Formula
+  desc "Official ODBC driver implementation for accessing ClickHouse as a data source"
+  homepage "https://github.com/ClickHouse/clickhouse-odbc#readme"
+  url "https://github.com/ClickHouse/clickhouse-odbc.git",
+    tag:      "v1.1.10.20210822",
+    revision: "c7aaff6860e448acee523f5f7d3ee97862fd07d2"
+  license "Apache-2.0"
+  head "https://github.com/ClickHouse/clickhouse-odbc.git",
+    branch:   "master"
+
+  depends_on "cmake" => :build
+  depends_on "pkg-config" => :build
+  depends_on "openssl@1.1"
+
+  uses_from_macos "icu4c"
+
+  on_macos do
+    depends_on "libiodbc"
+  end
+
+  on_linux do
+    depends_on "unixodbc"
+  end
+
+  def install
+    cmake_args = std_cmake_args.dup
+
+    cmake_args.map! { |x| x.start_with?("-DCMAKE_BUILD_TYPE=") ? "-DCMAKE_BUILD_TYPE=RelWithDebInfo" : x }
+    cmake_args << "-DCMAKE_BUILD_TYPE=RelWithDebInfo"
+
```

Comment:
Why so? In those packages, we recommend building `RelWithDebInfo` instead of `Release`. Some CMake code may not even handle `Release` configuration properly.

Reply:
We don't build with debug info, and build with `-O2` by default on Linux. If you really want to be explicit about the optimisation level, you can pass `-DCMAKE_C[XX]_FLAGS_RELEASE='-O2 -DNDEBUG'`.

There's also no point trying to pass debug flags, since `brew` strips out `-g` flags before they make it to the compiler anyway.

Reply:
Some logic may depend on `CMAKE_BUILD_TYPE` being equal to `RelWithDebInfo` by itself.

Removal of the flags is not good either, I would assume the package build system knows better how to build its stuff...

This is more relevant to me since I am planning on adding `clickhouse-client` and `clickhouse-server` which are products built from a huge monorepo, with a complicated CMake config, static linking of third parties and some system runtime, etc. Brew invading these boundaries may cause problems...

Reply:
How else does the build change (other than the compiler flags in `CMAKE_C[XX]_FLAGS_{RELEASE,RELWITHDEBINFO}`) when we use `RelWithDebInfo` instead of `Release`?

Using this would make the formula quite misleading, since it'll say we're building debug info, but none will actually be included in the binary package.

Reply:
I checked, and it looks like I can leave `Release` for this formula. However, another formula I am working on (`clickhouse`), will definitely require `RelWithDebInfo` specifically, and `-g` is actually being passed there when I install the formula and build it from sources. The debug info (of a carefully controlled verbosity) will be baked into the binaries, so it won't be separate files to install with the package. Those are mainly symbols and other info required for meaningful stack trace generation.

Reply:
> `-g` is actually being passed there when I install the formula and build it from sources.

How did you verify this? Do you still have the build logs? They're usually in `~/Library/Logs/Homebrew` (macOS) or `~/.cache/Homebrew/Logs`.

Reply:
I see the compilation commands, probably the effect of `--verbose`, which I specify.

Reply:
Those are the commands that the build system emits before `brew` is done filtering them, so it doesn't reflect what the compiler actually sees. `-g` flags will get stripped here: https://github.com/Homebrew/brew/blob/7e620666fb4d3165d057511f6ef3e7b37b38cd6d/Library/Homebrew/shims/super/cc#L179

You should see this reflected in the build logs I asked about above.

Reply:
Yeah, looks good as far as I can tell. What's the point in messing with the package's configuration from brew side?
```
clang++ called with: -fdiagnostics-color=always -fsized-deallocation -gdwarf-aranges -pipe -msse4.1 -msse4.2 -mpopcnt -fasynchronous-unwind-tables -falign-functions=32 -Wall -Wno-unused-command-line-argument -stdlib=libc++ -fdiagnostics-absolute-paths -fexperimental-new-pass-manager -Werror -Wextra -Wpedantic -Wno-vla-extension -Wno-zero-length-array -Wno-c11-extensions -Wcomma -Wconditional-uninitialized -Wcovered-switch-default -Wdeprecated -Wembedded-directive -Wempty-init-stmt -Wextra-semi-stmt -Wextra-semi -Wgnu-case-range -Winconsistent-missing-destructor-override -Wnewline-eof -Wold-style-cast -Wrange-loop-analysis -Wredundant-parens -Wreserved-id-macro -Wshadow-field -Wshadow-uncaptured-local -Wshadow -Wstring-plus-int -Wundef -Wunreachable-code-return -Wunreachable-code -Wunused-exception-parameter -Wunused-macros -Wunused-member-function -Wzero-as-null-pointer-constant -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c99-extensions -Wno-conversion -Wno-ctad-maybe-unsupported -Wno-deprecated-dynamic-exception-spec -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-nested-anon-types -Wno-packed -Wno-padded -Wno-return-std-move-in-c++11 -Wno-shift-sign-overflow -Wno-sign-conversion -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -O2 -g -DNDEBUG -O3 -flto=thin -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=10.15 -Wl,-search_paths_first -Wl,-headerpad_max_install_names -rdynamic -Wl,-U,_inside_main -flto=thin src/CMakeFiles/clickhouse_malloc.dir/Common/malloc.cpp.o utils/convert-month-partitioned-parts/CMakeFiles/convert-month-partitioned-parts.dir/main.cpp.o -o utils/convert-month-partitioned-parts/convert-month-partitioned-parts src/libclickhouse_new_delete.a src/AggregateFunctions/libclickhouse_aggregate_functions.a src/libdbms.a src/Parsers/libclickhouse_parsers.a contrib/boost-cmake/lib_boost_program_options.a contrib/jemalloc-cmake/libjemalloc.a src/Parsers/New/libclickhouse_parsers_new.a src/Parsers/libclickhouse_parsers.a contrib/antlr4-runtime-cmake/libantlr4-runtime.a contrib/cppkafka-cmake/libcppkafka.a contrib/librdkafka-cmake/librdkafka.a contrib/cyrus-sasl-cmake/libsasl2.a contrib/krb5-cmake/libkrb5.a contrib/nuraft-cmake/libnuraft.a contrib/boost-cmake/lib_boost_coroutine.a src/Common/Config/libclickhouse_common_config.a contrib/yaml-cpp-cmake/libyaml-cpp.a src/Common/ZooKeeper/libclickhouse_common_zookeeper.a src/Dictionaries/Embedded/libclickhouse_dictionaries_embedded.a src/libclickhouse_common_io.a contrib/boost-cmake/lib_boost_program_options.a base/widechar_width/libwidechar_width.a base/common/libcommon.a contrib/poco-cmake/Net/SSL/lib_poco_net_ssl.a contrib/poco-cmake/Crypto/lib_poco_crypto.a contrib/poco-cmake/Util/lib_poco_util.a contrib/poco-cmake/XML/lib_poco_xml.a contrib/poco-cmake/XML/lib_poco_xml_expat.a contrib/replxx-cmake/libreplxx.a contrib/cctz-cmake/libcctz.a -Wl,-force_load /tmp/clickhouse-20210922-42940-1aij01q/ClickHouse/build/contrib/cctz-cmake/libtzdata.a -Wl, contrib/fmtlib-cmake/libfmt.a contrib/dragonbox-cmake/libdragonbox_to_chars.a contrib/re2_st/libre2_st.a contrib/libcpuid-cmake/libcpuid.a contrib/croaring-cmake/libroaring.a contrib/xz/liblzma.a contrib/aws-s3-cmake/libaws_s3.a contrib/aws-s3-cmake/libaws_s3_checksums.a contrib/brotli-cmake/libbrotli.a contrib/poco-cmake/JSON/lib_poco_json.a contrib/poco-cmake/JSON/lib_poco_json_pdjson.a contrib/poco-cmake/MongoDB/lib_poco_mongodb.a contrib/poco-cmake/Net/lib_poco_net.a contrib/poco-cmake/Foundation/lib_poco_foundation.a contrib/poco-cmake/Foundation/lib_poco_foundation_pcre.a src/Common/StringUtils/libstring_utils.a contrib/boost-cmake/lib_boost_system.a contrib/capnproto-cmake/libcapnpc.a contrib/capnproto-cmake/libcapnp.a contrib/capnproto-cmake/libkj.a contrib/arrow-cmake/libparquet_static.a contrib/arrow-cmake/libarrow_static.a contrib/boost-cmake/lib_boost_filesystem.a contrib/double-conversion-cmake/libdouble-conversion.a contrib/flatbuffers/libflatbuffers.a contrib/arrow-cmake/libthrift_static.a contrib/boost-cmake/lib_boost_regex.a contrib/avro-cmake/libavrocpp.a contrib/boost-cmake/lib_boost_iostreams.a contrib/openldap-cmake/libldap_r.a contrib/openldap-cmake/liblber.a src/Server/grpc_protos/libclickhouse_grpc_protos.a contrib/grpc/libgrpc++.a contrib/protobuf/libprotobuf.a contrib/grpc/libgrpc.a contrib/re2/libre2.a contrib/grpc/third_party/cares/cares/lib/libcares.a -lresolv contrib/abseil-cpp/absl/status/libabsl_status.a contrib/abseil-cpp/absl/strings/libabsl_cord.a contrib/abseil-cpp/absl/hash/libabsl_hash.a contrib/abseil-cpp/absl/types/libabsl_bad_variant_access.a contrib/abseil-cpp/absl/hash/libabsl_city.a contrib/abseil-cpp/absl/container/libabsl_raw_hash_set.a contrib/abseil-cpp/absl/types/libabsl_bad_optional_access.a contrib/abseil-cpp/absl/container/libabsl_hashtablez_sampler.a contrib/abseil-cpp/absl/base/libabsl_exponential_biased.a -framework CoreFoundation contrib/grpc/libaddress_sorting.a contrib/grpc/libupb.a contrib/grpc/libgpr.a contrib/abseil-cpp/absl/synchronization/libabsl_synchronization.a contrib/abseil-cpp/absl/debugging/libabsl_stacktrace.a contrib/abseil-cpp/absl/debugging/libabsl_symbolize.a contrib/abseil-cpp/absl/debugging/libabsl_debugging_internal.a contrib/abseil-cpp/absl/debugging/libabsl_demangle_internal.a contrib/abseil-cpp/absl/synchronization/libabsl_graphcycles_internal.a contrib/abseil-cpp/absl/time/libabsl_time.a contrib/abseil-cpp/absl/time/libabsl_civil_time.a contrib/abseil-cpp/absl/time/libabsl_time_zone.a -framework CoreFoundation contrib/abseil-cpp/absl/base/libabsl_malloc_internal.a contrib/abseil-cpp/absl/strings/libabsl_str_format_internal.a contrib/abseil-cpp/absl/strings/libabsl_strings.a contrib/abseil-cpp/absl/strings/libabsl_strings_internal.a contrib/abseil-cpp/absl/numeric/libabsl_int128.a contrib/abseil-cpp/absl/base/libabsl_throw_delegate.a contrib/abseil-cpp/absl/base/libabsl_base.a contrib/abseil-cpp/absl/base/libabsl_raw_logging_internal.a contrib/abseil-cpp/absl/base/libabsl_dynamic_annotations.a contrib/abseil-cpp/absl/base/libabsl_log_severity.a contrib/abseil-cpp/absl/base/libabsl_spinlock_wait.a contrib/amqpcpp-cmake/libamqp-cpp.a libcassandra_static.a contrib/libuv/libuv_a.a contrib/rocksdb-cmake/librocksdb.a contrib/lz4-cmake/liblz4.a contrib/zstd-cmake/libzstd.a contrib/zlib-ng/libz.a contrib/snappy/libsnappy.a contrib/libpqxx-cmake/liblibpqxx.a contrib/libpq-cmake/liblibpq.a contrib/boringssl-cmake/libssl.a contrib/boringssl-cmake/libcrypto.a -lpthread contrib/boost-cmake/lib_boost_context.a contrib/cityhash102/libcityhash.a contrib/libcxx-cmake/libcxx.a contrib/libcxxabi-cmake/libcxxabi.a -nodefaultlibs -lc -lm -lpthread -ldl
superenv added:    -isysroot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk --sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -isystem/usr/local/include -isystem/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/binutils/include -I/usr/local/opt/llvm/include -L/usr/local/opt/binutils/lib -L/usr/local/opt/llvm/lib -L/usr/local/lib -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
superenv executed: clang++ -fdiagnostics-color=always -fsized-deallocation -gdwarf-aranges -pipe -msse4.1 -msse4.2 -mpopcnt -fasynchronous-unwind-tables -falign-functions=32 -Wall -Wno-unused-command-line-argument -stdlib=libc++ -fdiagnostics-absolute-paths -fexperimental-new-pass-manager -Werror -Wextra -Wpedantic -Wno-vla-extension -Wno-zero-length-array -Wno-c11-extensions -Wcomma -Wconditional-uninitialized -Wcovered-switch-default -Wdeprecated -Wembedded-directive -Wempty-init-stmt -Wextra-semi-stmt -Wextra-semi -Wgnu-case-range -Winconsistent-missing-destructor-override -Wnewline-eof -Wold-style-cast -Wrange-loop-analysis -Wredundant-parens -Wreserved-id-macro -Wshadow-field -Wshadow-uncaptured-local -Wshadow -Wstring-plus-int -Wundef -Wunreachable-code-return -Wunreachable-code -Wunused-exception-parameter -Wunused-macros -Wunused-member-function -Wzero-as-null-pointer-constant -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c99-extensions -Wno-conversion -Wno-ctad-maybe-unsupported -Wno-deprecated-dynamic-exception-spec -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-nested-anon-types -Wno-packed -Wno-padded -Wno-return-std-move-in-c++11 -Wno-shift-sign-overflow -Wno-sign-conversion -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -O2 -g -DNDEBUG -O3 -flto=thin -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=10.15 -Wl,-search_paths_first -Wl,-headerpad_max_install_names -rdynamic -Wl,-U,_inside_main -flto=thin src/CMakeFiles/clickhouse_malloc.dir/Common/malloc.cpp.o utils/convert-month-partitioned-parts/CMakeFiles/convert-month-partitioned-parts.dir/main.cpp.o -o utils/convert-month-partitioned-parts/convert-month-partitioned-parts src/libclickhouse_new_delete.a src/AggregateFunctions/libclickhouse_aggregate_functions.a src/libdbms.a src/Parsers/libclickhouse_parsers.a contrib/boost-cmake/lib_boost_program_options.a contrib/jemalloc-cmake/libjemalloc.a src/Parsers/New/libclickhouse_parsers_new.a src/Parsers/libclickhouse_parsers.a contrib/antlr4-runtime-cmake/libantlr4-runtime.a contrib/cppkafka-cmake/libcppkafka.a contrib/librdkafka-cmake/librdkafka.a contrib/cyrus-sasl-cmake/libsasl2.a contrib/krb5-cmake/libkrb5.a contrib/nuraft-cmake/libnuraft.a contrib/boost-cmake/lib_boost_coroutine.a src/Common/Config/libclickhouse_common_config.a contrib/yaml-cpp-cmake/libyaml-cpp.a src/Common/ZooKeeper/libclickhouse_common_zookeeper.a src/Dictionaries/Embedded/libclickhouse_dictionaries_embedded.a src/libclickhouse_common_io.a contrib/boost-cmake/lib_boost_program_options.a base/widechar_width/libwidechar_width.a base/common/libcommon.a contrib/poco-cmake/Net/SSL/lib_poco_net_ssl.a contrib/poco-cmake/Crypto/lib_poco_crypto.a contrib/poco-cmake/Util/lib_poco_util.a contrib/poco-cmake/XML/lib_poco_xml.a contrib/poco-cmake/XML/lib_poco_xml_expat.a contrib/replxx-cmake/libreplxx.a contrib/cctz-cmake/libcctz.a -Wl,-force_load /tmp/clickhouse-20210922-42940-1aij01q/ClickHouse/build/contrib/cctz-cmake/libtzdata.a -Wl, contrib/fmtlib-cmake/libfmt.a contrib/dragonbox-cmake/libdragonbox_to_chars.a contrib/re2_st/libre2_st.a contrib/libcpuid-cmake/libcpuid.a contrib/croaring-cmake/libroaring.a contrib/xz/liblzma.a contrib/aws-s3-cmake/libaws_s3.a contrib/aws-s3-cmake/libaws_s3_checksums.a contrib/brotli-cmake/libbrotli.a contrib/poco-cmake/JSON/lib_poco_json.a contrib/poco-cmake/JSON/lib_poco_json_pdjson.a contrib/poco-cmake/MongoDB/lib_poco_mongodb.a contrib/poco-cmake/Net/lib_poco_net.a contrib/poco-cmake/Foundation/lib_poco_foundation.a contrib/poco-cmake/Foundation/lib_poco_foundation_pcre.a src/Common/StringUtils/libstring_utils.a contrib/boost-cmake/lib_boost_system.a contrib/capnproto-cmake/libcapnpc.a contrib/capnproto-cmake/libcapnp.a contrib/capnproto-cmake/libkj.a contrib/arrow-cmake/libparquet_static.a contrib/arrow-cmake/libarrow_static.a contrib/boost-cmake/lib_boost_filesystem.a contrib/double-conversion-cmake/libdouble-conversion.a contrib/flatbuffers/libflatbuffers.a contrib/arrow-cmake/libthrift_static.a contrib/boost-cmake/lib_boost_regex.a contrib/avro-cmake/libavrocpp.a contrib/boost-cmake/lib_boost_iostreams.a contrib/openldap-cmake/libldap_r.a contrib/openldap-cmake/liblber.a src/Server/grpc_protos/libclickhouse_grpc_protos.a contrib/grpc/libgrpc++.a contrib/protobuf/libprotobuf.a contrib/grpc/libgrpc.a contrib/re2/libre2.a contrib/grpc/third_party/cares/cares/lib/libcares.a -lresolv contrib/abseil-cpp/absl/status/libabsl_status.a contrib/abseil-cpp/absl/strings/libabsl_cord.a contrib/abseil-cpp/absl/hash/libabsl_hash.a contrib/abseil-cpp/absl/types/libabsl_bad_variant_access.a contrib/abseil-cpp/absl/hash/libabsl_city.a contrib/abseil-cpp/absl/container/libabsl_raw_hash_set.a contrib/abseil-cpp/absl/types/libabsl_bad_optional_access.a contrib/abseil-cpp/absl/container/libabsl_hashtablez_sampler.a contrib/abseil-cpp/absl/base/libabsl_exponential_biased.a -framework CoreFoundation contrib/grpc/libaddress_sorting.a contrib/grpc/libupb.a contrib/grpc/libgpr.a contrib/abseil-cpp/absl/synchronization/libabsl_synchronization.a contrib/abseil-cpp/absl/debugging/libabsl_stacktrace.a contrib/abseil-cpp/absl/debugging/libabsl_symbolize.a contrib/abseil-cpp/absl/debugging/libabsl_debugging_internal.a contrib/abseil-cpp/absl/debugging/libabsl_demangle_internal.a contrib/abseil-cpp/absl/synchronization/libabsl_graphcycles_internal.a contrib/abseil-cpp/absl/time/libabsl_time.a contrib/abseil-cpp/absl/time/libabsl_civil_time.a contrib/abseil-cpp/absl/time/libabsl_time_zone.a -framework CoreFoundation contrib/abseil-cpp/absl/base/libabsl_malloc_internal.a contrib/abseil-cpp/absl/strings/libabsl_str_format_internal.a contrib/abseil-cpp/absl/strings/libabsl_strings.a contrib/abseil-cpp/absl/strings/libabsl_strings_internal.a contrib/abseil-cpp/absl/numeric/libabsl_int128.a contrib/abseil-cpp/absl/base/libabsl_throw_delegate.a contrib/abseil-cpp/absl/base/libabsl_base.a contrib/abseil-cpp/absl/base/libabsl_raw_logging_internal.a contrib/abseil-cpp/absl/base/libabsl_dynamic_annotations.a contrib/abseil-cpp/absl/base/libabsl_log_severity.a contrib/abseil-cpp/absl/base/libabsl_spinlock_wait.a contrib/amqpcpp-cmake/libamqp-cpp.a libcassandra_static.a contrib/libuv/libuv_a.a contrib/rocksdb-cmake/librocksdb.a contrib/lz4-cmake/liblz4.a contrib/zstd-cmake/libzstd.a contrib/zlib-ng/libz.a contrib/snappy/libsnappy.a contrib/libpqxx-cmake/liblibpqxx.a contrib/libpq-cmake/liblibpq.a contrib/boringssl-cmake/libssl.a contrib/boringssl-cmake/libcrypto.a -lpthread contrib/boost-cmake/lib_boost_context.a contrib/cityhash102/libcityhash.a contrib/libcxx-cmake/libcxx.a contrib/libcxxabi-cmake/libcxxabi.a -nodefaultlibs -lc -lm -lpthread -ldl -isysroot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk --sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -isystem/usr/local/include -isystem/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/binutils/include -I/usr/local/opt/llvm/include -L/usr/local/opt/binutils/lib -L/usr/local/opt/llvm/lib -L/usr/local/lib -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries -Wl,-headerpad_max_install_names
```

Reply:
There are many reasons, all of them supported by nearly a decade of experience building thousands of packages. In the case of debug info in particular: most users don't need or want them, and they often break the relocatability of binary packages.

Note that `-gdwarf` not being filtered is an implementation detail -- you might be able to use this for now, but there's no guarantee about this going away at some point in the future.